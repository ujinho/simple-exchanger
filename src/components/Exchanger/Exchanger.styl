$mediaBreakPointSmall = 767px

.exchanger
  padding 12px
  width 720px
  display flex
  flex-flow column
  justify-content space-between
  margin 24px auto
  text-align center

  &-hint
    height 24px

  &-pockets
    display flex
    justify-content space-around
    min-height 80px

  &-input-row
    font-size 24px
    line-height 1.47

  &-input
    width 180px
    margin-right 12px
    font-size 24px
    line-height 1.47
    text-align right

  &-buttons
    margin 24px 0 0
    display flex
    flex-direction column
    align-items center

  &-button
    width 240px
    font-size 24px
    line-height 1.47
    color #eee
    background-color #3a82c6
    cursor pointer

    & + &
      margin-top 8px

    &_outlined
      background-color white
      color #3a82c6
      outline 0
      border 0
      font-size 14px
      font-weight 200

    &_disabled
      cursor none
      pointer-events none
      opacity .2

  @media screen and (max-width: $mediaBreakPointSmall)
    width 100%
    padding 0

    &-input-row
      margin-top 24px

    &-pockets
      flex-flow column wrap
      align-items center
      min-height 240px

    &-button
      margin-top 12px

    &-input
      width 154px
      padding 0 12px
      border 1px solid lightgray
