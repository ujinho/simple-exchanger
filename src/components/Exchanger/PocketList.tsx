import React, { memo } from 'react';
import Pocket from 'components/Pocket';
import { DEFAULT_CURRENCY_FORMAT } from 'const';
import { cn } from '@bem-react/classname';
import { Currency } from 'dinero.js';

type Props = {
  from: Currency;
  to: Currency;
  subtrahend: Dinero.Dinero;
  converted: Dinero.Dinero;
  pockets: ReadonlyArray<Dinero.Dinero>;
  rates: { [key in Currency]?: number };
  onSelectPocket: (symbol: Currency) => void;
};

const exchangerCn = cn('exchanger');

const PocketList = ({ from, to, rates, subtrahend, converted, pockets, onSelectPocket }: Props) => (
  <div className={exchangerCn('pockets')}>
    {pockets.map(pocket => {
      const currency = pocket.getCurrency() as Currency;
      const amountFormatted = pocket.toFormat(DEFAULT_CURRENCY_FORMAT);
      const asSource = currency === from;
      const asDestination = currency === to;
      let convertedAmount: number;

      if (asSource) {
        convertedAmount = subtrahend && subtrahend.multiply(-1).getAmount();
      } else if (asDestination) {
        convertedAmount = converted && converted.getAmount();
      }

      return (
        <Pocket
          key={currency}
          asSource={asSource}
          asDestination={asDestination}
          currency={currency}
          amount={amountFormatted}
          rate={rates && rates[currency]}
          converted={convertedAmount}
          onSelect={onSelectPocket}
        />
      );
    })}
  </div>
);

export default memo(PocketList);
