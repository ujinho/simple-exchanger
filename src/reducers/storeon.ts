import storeon from 'storeon/react';

export const useStoreon = storeon;
