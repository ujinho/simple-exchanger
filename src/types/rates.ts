import { Currency } from 'dinero.js';

export interface FxRates {
  base: Currency;
  date?: string;
  rates: { [key in Currency]?: number };
  error?: string;
}
