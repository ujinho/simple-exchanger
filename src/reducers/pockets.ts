import Dinero, { Currency } from 'dinero.js';
import { StoreonEvents, Module, Store } from 'storeon';

type PocketsMap = { [key in Currency]?: Dinero.Dinero };

export interface PocketsState {
  to: Currency;
  from: Currency;
  amount: number;
  pockets: PocketsMap;
  converted: Dinero.Dinero;
  subtrahend: Dinero.Dinero;
  error: string;
}

export type ConvertPayload = {
  amount: number;
  from: Currency;
  to: Currency;
};

export enum PocketsEventName {
  Set = 'pockets/set',
  Reset = 'pockets/reset',
  PreConvert = 'pockets/pre-convert',
  Convert = 'pockets/convert',
}

export interface PocketsEvents extends StoreonEvents<PocketsState> {
  [PocketsEventName.Set]: Partial<PocketsState>;
  [PocketsEventName.Reset]: Partial<PocketsState>;
  [PocketsEventName.Convert]: ConvertPayload;
  [PocketsEventName.PreConvert]: ConvertPayload;
}

export type PocketsStore = Store<PocketsState, PocketsEvents>;

type ConvertReducer<R> = (state: PocketsState, payload: ConvertPayload) => Promise<R> & R;

type ConvertReducerCreator = (store: PocketsStore) => ConvertReducer<Partial<PocketsState>>;

const pocketsInitial: PocketsMap = {
  USD: Dinero({ amount: 50000, currency: 'USD' }),
  EUR: Dinero({ amount: 50000, currency: 'EUR' }),
  GBP: Dinero({ amount: 50000, currency: 'GBP' }),
};

export const initialState: PocketsState = {
  to: undefined,
  from: undefined,
  amount: 0,
  converted: undefined,
  subtrahend: undefined,
  error: undefined,
  pockets: pocketsInitial,
};

export const getPreConvertReducer: ConvertReducerCreator = store => (
  state,
  { amount, from, to },
) => {
  if (amount == null) {
    return Promise.resolve({});
  }

  if (from === to) {
    const stateChanges = { amount };
    store.dispatch(PocketsEventName.Set, stateChanges);
    return Promise.resolve(stateChanges);
  }

  const pocketFrom = state.pockets[from];
  const pocketTo = state.pockets[to];

  if (!pocketFrom || !pocketTo) {
    const stateChanges = { amount };
    store.dispatch(PocketsEventName.Set, { amount });
    return Promise.resolve(stateChanges);
  }

  const subtrahend = Dinero({ currency: from, amount });
  return subtrahend
    .convert(to)
    .then(converted => {
      const stateChanges = {
        to,
        from,
        amount,
        converted,
        subtrahend,
      };
      store.dispatch(PocketsEventName.Set, stateChanges);
      return Promise.resolve(stateChanges);
    })
    .catch(e => {
      /* error handling should be placed here */
      return Promise.resolve({
        error: e.message,
      });
    });
};

export const getConvertReducer: ConvertReducerCreator = store => (state, { amount, from, to }) => {
  if (!from || !to || !amount) {
    return Promise.resolve({});
  }

  if (from === to) {
    return Promise.resolve({});
  }

  const pocketFrom = state.pockets[from];
  const pocketTo = state.pockets[to];

  if (!pocketFrom || !pocketTo) {
    return Promise.resolve({});
  }

  const subtrahend = Dinero({ currency: from, amount });
  const substracted = pocketFrom && pocketFrom.subtract(subtrahend);

  if (!substracted) return Promise.resolve({});

  return subtrahend
    .convert(to)
    .then(converted => {
      const stateChanges = {
        to,
        from,
        amount,
        pockets: {
          ...state.pockets,
          [from]: substracted,
          [to]: pocketTo.add(converted),
        },
      };
      store.dispatch(PocketsEventName.Set, stateChanges);
      return Promise.resolve(stateChanges);
    })
    .catch(e => {
      return Promise.resolve({
        error: e.message,
      });
    });
};

const pocketsModule: Module<PocketsState, PocketsEvents> = store => {
  store.on('@init', () => initialState);

  store.on(PocketsEventName.Reset, (_1, _2) => {
    return { pockets: pocketsInitial };
  });

  store.on(PocketsEventName.Set, (_, stateChanges) => {
    return stateChanges;
  });

  store.on(PocketsEventName.PreConvert, getPreConvertReducer(store));

  store.on(PocketsEventName.Convert, getConvertReducer(store));
};

export default pocketsModule;
