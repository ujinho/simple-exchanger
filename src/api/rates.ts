import Axios from 'axios';
import { Currency } from 'dinero.js';
import { FxRates } from 'types/rates';

export const axios = Axios.create({
  baseURL: 'https://api.ratesapi.io/api/',
  timeout: 10000,
  headers: { Accept: 'application/json' },
});

export const getLatestRates = (base: Currency, symbolsArray: ReadonlyArray<Currency>) => {
  const symbols = (symbolsArray && symbolsArray.join(',')) || '';
  return axios.get<FxRates>('latest', {
    params: {
      base,
      symbols,
    },
  });
};
