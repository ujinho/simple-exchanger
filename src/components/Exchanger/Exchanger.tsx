import React, { useCallback, useMemo, useEffect } from 'react';

import { useStoreon } from 'reducers/storeon';

import Dinero, { Currency } from 'dinero.js';
import { cn } from '@bem-react/classname';
import { SYMBOLS } from 'const';

import PocketList from './PocketList';

import { AppState } from '../../store';

import './Exchanger.styl';
import { PocketsEventName } from 'reducers/pockets';
import { RatesEventName } from 'reducers/rates';

const exchangerClassName = cn('exchanger');

export default () => {
  const { dispatch, from, to, amount, rates, pockets, subtrahend, converted } = useStoreon<
    AppState
  >('from', 'to', 'amount', 'rates', 'pockets', 'subtrahend', 'converted');

  const isConvertDisabled = amount > pockets[from]?.getAmount() || !from || !to || !amount;

  useEffect(() => {
    const symbols = SYMBOLS.filter(sym => sym !== from);
    dispatch(RatesEventName.Poll, { base: from, symbols });
  }, [from]);

  useEffect(() => {
    // @ts-ignore
    Dinero.globalExchangeRatesApi?.endpoint = new Promise(resolve => resolve(rates));
  }, [rates]);

  const onSelectPocket = useCallback(
    (currency: Currency) => {
      if (to && to === currency) {
        dispatch(PocketsEventName.Set, {
          from: currency,
          to: undefined,
        });
      } else if (from && from !== currency) {
        dispatch(PocketsEventName.Set, { to: currency });
      } else {
        dispatch(PocketsEventName.Set, { from: currency });
      }
    },
    [from, to],
  );

  useEffect(() => {
    if (!from || !to || !amount) {
      return;
    }
    if (!isConvertDisabled) {
      dispatch(PocketsEventName.PreConvert, { from, to, amount });
    } else {
      dispatch(PocketsEventName.Set, { subtrahend: null, converted: null });
    }
  }, [from, to, amount, isConvertDisabled]);

  const onSetAmount = useCallback(e => {
    const value = Math.round(parseFloat(e.currentTarget.value) * 10000) / 100;
    dispatch(PocketsEventName.Set, { amount: value });
  }, []);

  const onConvert = useCallback(() => {
    dispatch(PocketsEventName.Convert, { from, to, amount });
  }, [from, to, amount]);

  const onReset = useCallback(() => {
    dispatch(PocketsEventName.Reset);
  }, []);

  const pocketsAsArray: ReadonlyArray<Dinero.Dinero> = useMemo(() => Object.values(pockets), [
    pockets,
  ]);

  const max = pockets[from]?.divide(100).getAmount() || 500;

  return (
    <div className={exchangerClassName()}>
      <div data-testid="exchanger-hint" className={exchangerClassName('hint')}>
        {!from && 'Select source pocket'}
        {from && !to && 'Select destination pocket'}
        {from && to && !amount && 'Set amount to convert'}
      </div>

      <PocketList
        from={from}
        to={to}
        rates={rates?.rates}
        subtrahend={subtrahend}
        converted={converted}
        pockets={pocketsAsArray}
        data-testid="pocket-list"
        onSelectPocket={onSelectPocket}
      />

      <div className={exchangerClassName('input-row')}>
        <input
          min={-max}
          max={max}
          type="number"
          step="0.05"
          pattern="[0-9]+([,\.][0-9]+)?"
          placeholder="0.00"
          className={exchangerClassName('input')}
          data-testid="amount-input"
          onChange={onSetAmount}
        />
        {from}
      </div>

      <div className={exchangerClassName('buttons')}>
        <button
          className={exchangerClassName('button', { disabled: isConvertDisabled })}
          disabled={isConvertDisabled}
          data-testid="convert-button"
          onClick={onConvert}
        >
          Convert
        </button>

        <button
          className={exchangerClassName('button', { outlined: true })}
          data-testid="reset-button"
          onClick={onReset}
        >
          Reset
        </button>
      </div>
    </div>
  );
};
