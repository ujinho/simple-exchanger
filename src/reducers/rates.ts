import { getLatestRates } from 'api/rates';
import { POLL_INTERVAL } from 'const';
import { StoreonEvents, Module } from 'storeon';
import { Currency } from 'dinero.js';
import { FxRates } from 'types/rates';

export type RatesState = {
  base: Currency;
  symbols: ReadonlyArray<Currency>;
  pollInterval: number;
  rates: FxRates;
  intervalId: number;
};

type PollPayload = {
  base: Currency;
  symbols: ReadonlyArray<Currency>;
};

export enum RatesEventName {
  Set = 'rates/set',
  Poll = 'rates/poll',
  SetIntervalId = 'rates/set-interval-id',
}

export interface RatesEvents extends StoreonEvents<RatesState> {
  [RatesEventName.Set]: Partial<RatesState>;
  [RatesEventName.Poll]: PollPayload;
  [RatesEventName.SetIntervalId]: number;
}

const initialState: RatesState = {
  base: undefined,
  symbols: [],
  pollInterval: POLL_INTERVAL,
  rates: undefined,
  intervalId: undefined,
};

const ratesModule: Module<RatesState, RatesEvents> = store => {
  store.on('@init', () => initialState);

  store.on(RatesEventName.SetIntervalId, (_, intervalId) => ({
    intervalId,
  }));

  store.on(RatesEventName.Set, (_, nextState) => nextState);

  store.on(RatesEventName.Poll, (_, { base, symbols }) => {
    const { pollInterval, intervalId } = store.get();

    if (intervalId) {
      clearInterval(intervalId);
    }

    if (!base || !symbols) {
      return {};
    }

    const action = () => {
      return getLatestRates(base, symbols)
        .then(response => {
          const rates = response?.data;
          const nextState = {
            symbols,
            base,
            rates,
          };
          return store.dispatch(RatesEventName.Set, nextState);
        })
        .catch(() => {
          /* error handling should be here */
        });
    };

    action();
    const nextIntervalId = window.setInterval(action, pollInterval);
    store.dispatch(RatesEventName.SetIntervalId, nextIntervalId);
  });
};

export default ratesModule;
