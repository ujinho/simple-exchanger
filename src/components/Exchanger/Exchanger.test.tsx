import React from 'react';

import Dinero from 'dinero.js';

import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/react/cleanup-after-each';

import { mockUseStoreon } from 'utils/test';

import Exchanger from './Exchanger';
import { FxRates } from 'types/rates';
import { PocketsEventName } from 'reducers/pockets';
import { RatesEventName } from 'reducers/rates';

const INITIAL_AMOUNT = 500;
const INITIAL_POCKET_AMOUNT = 5000;

const dispatch = jest.fn();

const rates: FxRates = {
  base: 'GBP',
  rates: {
    EUR: 1.1,
    USD: 1.2,
  },
};

const from = 'GBP';
const to = 'USD';
const amount = INITIAL_AMOUNT;
const pockets = {
  USD: Dinero({ amount: INITIAL_POCKET_AMOUNT, currency: 'USD' }),
  EUR: Dinero({ amount: INITIAL_POCKET_AMOUNT, currency: 'EUR' }),
  GBP: Dinero({ amount: INITIAL_POCKET_AMOUNT, currency: 'GBP' }),
};
const subtrahend = Dinero({ currency: from, amount });
const converted = Dinero({ currency: from, amount: amount * 1.2 });

const defaultState = { dispatch, rates, to, from, amount, pockets, subtrahend, converted };

describe('<Exchanger /> spec', () => {
  beforeEach(() => {
    mockUseStoreon(defaultState);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render without crasing', () => {
    const element = render(<Exchanger />);
    expect(element).toMatchSnapshot();
  });

  it('convert button should be disabled if no source pocket selected', () => {
    mockUseStoreon({
      ...defaultState,
      from: undefined,
    });
    const element = render(<Exchanger />);
    expect(element.getByTestId('convert-button')).toBeDisabled();
  });

  it('convert button should be disabled if no destination pocket selected', () => {
    mockUseStoreon({
      ...defaultState,
      to: undefined,
    });

    const element = render(<Exchanger />);
    expect(element.getByTestId('convert-button')).toBeDisabled();
  });

  it('convert button should be disabled if amount is not specified', () => {
    mockUseStoreon({
      ...defaultState,
      amount: 0,
    });
    const element = render(<Exchanger />);
    expect(element.getByTestId('convert-button')).toBeDisabled();
  });

  it('convert button should be disabled if amount greater than source pocket amount', () => {
    mockUseStoreon({
      ...defaultState,
      amount: INITIAL_POCKET_AMOUNT + 1,
    });
    const element = render(<Exchanger />);
    expect(element.getByTestId('convert-button')).toBeDisabled();
  });

  it('should enable convert button when all fields are set', () => {
    mockUseStoreon({
      ...defaultState,
    });
    const element = render(<Exchanger />);
    expect(element.getByTestId('convert-button')).toBeEnabled();
  });

  it('should correctly render all pockets', () => {
    mockUseStoreon({
      ...defaultState,
    });
    const element = render(<Exchanger />);
    expect(element.getByTestId('pocket-USD')).toHaveTextContent('USD 50.00');
    expect(element.getByTestId('pocket-USD-converted-value')).toHaveTextContent('+6');
    expect(element.getByTestId('pocket-USD-rate')).toHaveTextContent('1.2');

    expect(element.getByTestId('pocket-GBP')).toHaveTextContent('GBP 50.00');
    expect(element.getByTestId('pocket-GBP-converted-value')).toHaveTextContent('-5');
    expect(element.getByTestId('pocket-GBP-rate')).toHaveTextContent('1.00');
  });

  it('should pre-convert to display substrahend and converted values', () => {
    render(<Exchanger />);

    expect(dispatch).toHaveBeenLastCalledWith(PocketsEventName.PreConvert, {
      amount,
      from,
      to,
    });
  });

  it('should save proper amount value to store', () => {
    const element = render(<Exchanger />);

    fireEvent.change(element.getByTestId('amount-input'), {
      target: {
        value: 11.5612,
      },
    });

    expect(dispatch).toHaveBeenLastCalledWith(PocketsEventName.Set, {
      amount: 1156.12,
    });
  });

  it('should convert with proper values', () => {
    mockUseStoreon({ ...defaultState, amount: 400, from: 'EUR', to: 'USD' });

    const element = render(<Exchanger />);

    fireEvent.click(element.getByTestId('convert-button'));

    expect(dispatch).toHaveBeenLastCalledWith(PocketsEventName.Convert, {
      amount: 400,
      from: 'EUR',
      to: 'USD',
    });
  });

  it('should be able to reset values', () => {
    mockUseStoreon({ ...defaultState, amount: 400, from: 'EUR', to: 'USD' });

    const element = render(<Exchanger />);

    fireEvent.click(element.getByTestId('reset-button'));

    expect(dispatch).toHaveBeenLastCalledWith(PocketsEventName.Reset);
  });

  it('should poll rates', () => {
    render(<Exchanger />);

    expect(dispatch).toHaveBeenCalledWith(RatesEventName.Poll, {
      base: 'GBP',
      symbols: expect.arrayContaining(['USD', 'EUR']),
    });
  });
});
