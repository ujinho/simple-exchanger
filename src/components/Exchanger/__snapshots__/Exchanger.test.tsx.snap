// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exchanger /> spec should render without crasing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="exchanger"
      >
        <div
          class="exchanger-hint"
          data-testid="exchanger-hint"
        />
        <div
          class="exchanger-pockets"
        >
          <div
            class="pocket pocket_destination"
            data-testid="pocket-USD"
          >
            USD 50.00
            <div
              class="pocket-info"
            >
              <div
                class="pocket-converted pocket-converted_sign_plus"
                data-testid="pocket-USD-converted-value"
              >
                +6
              </div>
              <div
                class="pocket-rate"
                data-testid="pocket-USD-rate"
              >
                1.2
              </div>
            </div>
          </div>
          <div
            class="pocket"
            data-testid="pocket-EUR"
          >
            EUR 50.00
            <div
              class="pocket-info"
            >
              <div
                class="pocket-converted pocket-converted_sign_minus"
                data-testid="pocket-EUR-converted-value"
              />
              <div
                class="pocket-rate"
                data-testid="pocket-EUR-rate"
              >
                1.1
              </div>
            </div>
          </div>
          <div
            class="pocket pocket_source"
            data-testid="pocket-GBP"
          >
            GBP 50.00
            <div
              class="pocket-info"
            >
              <div
                class="pocket-converted pocket-converted_sign_minus"
                data-testid="pocket-GBP-converted-value"
              >
                -5
              </div>
              <div
                class="pocket-rate"
                data-testid="pocket-GBP-rate"
              >
                1.00
              </div>
            </div>
          </div>
        </div>
        <div
          class="exchanger-input-row"
        >
          <input
            class="exchanger-input"
            data-testid="amount-input"
            max="50"
            min="-50"
            pattern="[0-9]+([,\\\\.][0-9]+)?"
            placeholder="0.00"
            step="0.05"
            type="number"
          />
          GBP
        </div>
        <div
          class="exchanger-buttons"
        >
          <button
            class="exchanger-button"
            data-testid="convert-button"
          >
            Convert
          </button>
          <button
            class="exchanger-button exchanger-button_outlined"
            data-testid="reset-button"
          >
            Reset
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="exchanger"
    >
      <div
        class="exchanger-hint"
        data-testid="exchanger-hint"
      />
      <div
        class="exchanger-pockets"
      >
        <div
          class="pocket pocket_destination"
          data-testid="pocket-USD"
        >
          USD 50.00
          <div
            class="pocket-info"
          >
            <div
              class="pocket-converted pocket-converted_sign_plus"
              data-testid="pocket-USD-converted-value"
            >
              +6
            </div>
            <div
              class="pocket-rate"
              data-testid="pocket-USD-rate"
            >
              1.2
            </div>
          </div>
        </div>
        <div
          class="pocket"
          data-testid="pocket-EUR"
        >
          EUR 50.00
          <div
            class="pocket-info"
          >
            <div
              class="pocket-converted pocket-converted_sign_minus"
              data-testid="pocket-EUR-converted-value"
            />
            <div
              class="pocket-rate"
              data-testid="pocket-EUR-rate"
            >
              1.1
            </div>
          </div>
        </div>
        <div
          class="pocket pocket_source"
          data-testid="pocket-GBP"
        >
          GBP 50.00
          <div
            class="pocket-info"
          >
            <div
              class="pocket-converted pocket-converted_sign_minus"
              data-testid="pocket-GBP-converted-value"
            >
              -5
            </div>
            <div
              class="pocket-rate"
              data-testid="pocket-GBP-rate"
            >
              1.00
            </div>
          </div>
        </div>
      </div>
      <div
        class="exchanger-input-row"
      >
        <input
          class="exchanger-input"
          data-testid="amount-input"
          max="50"
          min="-50"
          pattern="[0-9]+([,\\\\.][0-9]+)?"
          placeholder="0.00"
          step="0.05"
          type="number"
        />
        GBP
      </div>
      <div
        class="exchanger-buttons"
      >
        <button
          class="exchanger-button"
          data-testid="convert-button"
        >
          Convert
        </button>
        <button
          class="exchanger-button exchanger-button_outlined"
          data-testid="reset-button"
        >
          Reset
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
